### PR TITLE
Update for Python3

### PR DESCRIPTION
**Pull request by cairosanders**
Updates for python3 include
- syntax and library import updates
- Using mainline Pydap 3.2.2
- Using `pydap-extras` for PCIC handlers `RawPcicSqlHandler` and `ClimoPcicSqlHandler`

### CODE REVIEW

File: tests/test_util.py [hunk 1]
```diff
@@ -8,64 +8,100 @@
...
 
+
 def test_get_extension_good():
-    params = {'data-format': 'csv'}
-    req = Request.blank('?' + urlencode(params))
-    assert get_extension(req.environ) == 'csv'
+    params = {"data-format": "html"}
+    req = Request.blank("?" + urlencode(params))
+    assert get_extension(req.environ) == "html"
+
```
Comment (cairosanders) [inline tests/test_util.py:98]: This was failing, but I couldn't find csv response anywhere so I changed it to html. Perhaps there is something else I have to fix?

File: pdp_util/agg.py [hunk 2]
```diff
@@ -38,99 +40,126 @@ def __init__(self, dsn, sesh=None):
...
     :param stations: A list of (``network_name``, ``native_id``) pairs representing the stations for which this response should include variable metadata
     :param climo: Should these be climatological variables?
     :type climo: bool
     :rtype: iterator
-    '''
+    """
     for station in stations:
         network_name = station[0]
-        filename = '{0}/variables.csv'.format(network_name)
+        filename = "{0}/variables.csv".format(network_name)
```
Comment (nikola-rados) [inline pdp_util/agg.py:106]: f-string?

File: pdp_util/agg.py [hunk 3]
```diff
@@ -140,30 +169,37 @@ def get_pcds_responders(dsn, stns, extension, clip_dates, environ):
...
+        if "download-climatology" in form
+        else (RawPcicSqlHandler(dsn), "rsql")
+    )
 
     sdate, edate = clip_dates
     content_length = 0
     for net, stn in stns:
         newenv = environ.copy()
-        newenv['PATH_INFO'] = '/%s/%s.%s.%s' % (net, stn, handler_ext, extension)
+        newenv["PATH_INFO"] = "/%s/%s.%s.%s" % (net, stn, handler_ext, extension)
```
Comment (nikola-rados) [inline pdp_util/agg.py:186]: f-string?

File: pdp_util/agg.py [hunk 4]
```diff
@@ -140,30 +169,37 @@ def get_pcds_responders(dsn, stns, extension, clip_dates, environ):
...
-        if edate: qs.append("station_observations.time<='%s'" % edate)
-        newenv['QUERY_STRING'] = '&'.join(qs)
+        if sdate:
+            qs.append("station_observations.time>='%s'" % sdate)
+        if edate:
+            qs.append("station_observations.time<='%s'" % edate)
+        newenv["QUERY_STRING"] = "&".join(qs)
 
-        name = '%(net)s/%(stn)s.%(extension)s' % locals()
+        name = "%(net)s/%(stn)s.%(extension)s" % locals()
```
Comment (nikola-rados) [inline pdp_util/agg.py:195]: f-string?

File: pdp_util/dbdict.py [hunk 5]
```diff
@@ -22,19 +24,21 @@ def __len__(self):
...
-    defaults = {'dialect': 'postgresql', 'driver': 'psycopg2', 'sslmode': 'require'}
+    defaults = {"dialect": "postgresql", "driver": "psycopg2", "sslmode": "require"}
     defaults.update(d)
     d = defaults
-    if set(('database', 'user', 'host')) <= set(d):
-        d['login'] = '{user}:{password}'.format(**d) if 'password' in d else d['user']
-        return '{dialect}+{driver}://{login}@{host}/{database}'.format(**d)
+    if set(("database", "user", "host")) <= set(d):
+        d["login"] = "{user}:{password}".format(**d) if "password" in d else d["user"]
+        return "{dialect}+{driver}://{login}@{host}/{database}".format(**d)
```
Comment (nikola-rados) [inline pdp_util/dbdict.py:40]: f-string?

File: pdp_util/filters.py [hunk 6]
```diff
@@ -64,63 +69,60 @@ def validate(self, value):
         return value if m else None
 
     def __str__(self):
-        return '<FormFilter> %s:%s' % (self.name, self.value)
+        return "<FormFilter> %s:%s" % (self.name, self.value)
```
Comment (nikola-rados) [inline pdp_util/filters.py:72]: f-string?

File: pdp_util/filters.py [hunk 7]
```diff
@@ -64,63 +69,60 @@ def validate(self, value):
...
 
 
 def mk_mp_regex():
-    decimal = r'-?[0-9]+(.[0-9]+)?'
-    point = '%(decimal)s %(decimal)s' % locals()
-    inner = r'\(%(point)s(, ?%(point)s){2,}\)' % locals()
+    decimal = r"-?[0-9]+(.[0-9]+)?"
+    point = "%(decimal)s %(decimal)s" % locals()
+    inner = r"\(%(point)s(, ?%(point)s){2,}\)" % locals()
```
Comment (nikola-rados) [inline pdp_util/filters.py:78]: f-string?

File: pdp_util/filters.py [hunk 8]
```diff
@@ -64,63 +69,60 @@ def validate(self, value):
...
     outer = inner
-    polygon = r'\(%(outer)s(, ?%(inner)s)?\)' % locals()
-    multipolygon = r'MULTIPOLYGON ?\(%(polygon)s(, ?%(polygon)s)*\)' % locals()
+    polygon = r"\(%(outer)s(, ?%(inner)s)?\)" % locals()
+    multipolygon = r"MULTIPOLYGON ?\(%(polygon)s(, ?%(polygon)s)*\)" % locals()
```
Comment (nikola-rados) [inline pdp_util/filters.py:81]: f-string?

File: pdp_util/filters.py [hunk 9]
```diff
@@ -64,63 +69,60 @@ def validate(self, value):
...
+    "input-var": FormFilter(
         # Single variable identifier
         # See note in class docstring regarding the content of a variable
         # identifier.
-        'input-var',
-        r'[a-z: _]+',
-        lambda x: cng.vars.like('%{}%'.format(x))
+        "input-var",
+        r"[a-z: _]+",
+        lambda x: cng.vars.like("%{}%".format(x)),
```
Comment (nikola-rados) [inline pdp_util/filters.py:116]: f-string?

File: pdp_util/pcds_dispatch.py [hunk 10]
```diff
@@ -6,112 +6,139 @@
...
                 cls_ = RawPcicSqlHandler
 
             env = {}
-            if '.' not in path: # Assume that it's just a listing and not point to a dataset
-                env['PATH_INFO'] = path.rstrip('/') + '.%s.html' % ext #FIXME html response!
+            if (
+                "." not in path
+            ):  # Assume that it's just a listing and not point to a dataset
+                env["PATH_INFO"] = (
+                    path.rstrip("/") + ".%s.html" % ext
```
Comment (nikola-rados) [inline pdp_util/pcds_dispatch.py:134]: f-string?

**Review by nikola-rados — CHANGES_REQUESTED** [review]
There are a **ton** of places where we could be using f-strings here. I've labelled some of them below. Otherwise this looks pretty good!

File: pdp_util/filters.py (hunk 7 above)
Comment (cairosanders) [reply pdp_util/filters.py:78]: Can I do that with a regular expression?

Comment (nikola-rados) [reply pdp_util/filters.py:78]: Oh sorry I didn't notice some of these were regex.

File: tests/test_agg.py [hunk 11]
```diff
@@ -7,29 +7,39 @@
...
+# import pydap.handlers.pcic
 
 import pytest
+import pydap_extras
 
-stns = [('ARDA', '115084'), ('EC_raw', '1046332'), ('FLNRO-WMB', '369')]
+stns = [("ARDA", "115084"), ("EC_raw", "1046332"), ("FLNRO-WMB", "369")]
+
+# def test_can_instantiate(test_session, conn_params):
+#     app = PcdsZipApp(conn_params, test_session)
```
Comment (nikola-rados) [inline tests/test_agg.py:26]: Can this be removed if we are not using it?

**Review by nikola-rados — COMMENTED** [review]
There are a couple things that could use changing but overall this looks good. One comment I will add is that we may want to separate functional and formatting changes. I can be very hard to find the functional changes when they are combined.

File: pdp_util/counts.py [hunk 12]
```diff
@@ -63,45 +68,61 @@ def __call__(self, environ, start_response):
...
-if __name__ == '__main__':
+
+if __name__ == "__main__":
     port = 8555
     from flask import Flask
-    conn_params = {'database': 'crmp', 'user': 'hiebert', 'host': 'monsoon.pcic.uvic.ca'}
+
+    conn_params = {
+        "database": "crmp",
+        "user": "hiebert",
```
[the remark this answers is not in the document]
Comment (cairosanders) [reply pdp_util/counts.py:121]: I'm not sure. Maybe @jameshiebert will have an answer to that?

[the remark this answers is not in the document]
Comment (jameshiebert) [reply pdp_util/counts.py:121]: AFAIK, no one is actually running this module as a script, so we could probably axe the entire block after `if __name__ == '__main__':`.

File: .github/workflows/python-ci.yml [hunk 13]
```diff
@@ -20,8 +20,9 @@ jobs:
       env:
         PIP_INDEX_URL: https://pypi.pacificclimate.org/simple
       run: |
+        sudo apt-get update
         sudo apt purge postgresql-client-13 postgresql-server-dev-13
-        sudo apt-get install libhdf5-serial-dev libnetcdf-dev libspatialite-dev postgresql-9.5-postgis-2.4
+        sudo apt-get install libhdf5-serial-dev libnetcdf-dev libspatialite-dev postgresql-12-postgis-3
```
Comment (jameshiebert) [inline .github/workflows/python-ci.yml:25]: Based on my experience yesterday, I think that you need to `sudo apt purge postgresql-client-13 postgresql-server-dev-all`

Comment (cairosanders) [reply .github/workflows/python-ci.yml:25]: Thanks :pray:

**Comment by cairosanders** [comment]
@nikola-rados @jameshiebert do you think this is okay to merge?

**Comment by nikola-rados** [comment]
I would say it looks good!

File: pdp_util/agg.py [hunk 14]
```diff
@@ -38,99 +40,126 @@ def __init__(self, dsn, sesh=None):
...
-       :param responders: A list of (``name``, ``generator``) pairs where ``name`` is the filename to use in the zip archive and ``generator`` should yield all bytes for a single file.
-       :rtype: iterator
-    '''
-    with SpooledTemporaryFile(1024*1024*1024) as f:
-        yield 'PK' # Response headers aren't sent until the first chunk of data is sent.  Let's get this repsonse moving!
-        z = ZipFile(f, 'w', ZIP_DEFLATED)
+    :param responders: A list of (``name``, ``generator``) pairs where ``name`` is the filename to use in the zip archive and ``generator`` should yield all bytes for a single file.
+    :rtype: iterator
+    """
+    with SpooledTemporaryFile(1024 * 3) as f:
```
Comment (jameshiebert) [inline pdp_util/agg.py:81]: Oooops! I didn't catch this before you merged, but this change is not correct. I think you meant `1024 ** 3`. Mind fixing that quickly?

Comment (cairosanders) [reply pdp_util/agg.py:81]: Yes thanks!

**Pull request merged**